kotlinx-coroutines-core/common/test/channels/ChannelsTest.kt

```diff
@@ -21,14 +21,19 @@ class ChannelsTest: TestBase() {
 
     @Test
     fun testCloseWithMultipleWaiters() = runTest {
+        // 1. Coroutine #1 suspends on `receive()`
```

Comment:
The original order was fixed on purpose: https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/common/src/channels/AbstractChannel.kt#L306

Can we avoid changing that?

Reply:
I've changed the test according to this comment:
>  It's important to traverse list from right to left to avoid races with sender.

It is indeed important to do that from right to left. Yet, the test assumed the opposite order before the fix.

Reply:
Wow, that was unclear. Should we also invoke the `onUndeliveredElement` lambda in the order of elements addition? The suggested implementation invokes the lambda in the reverse order, but no test fails.

Reply:
>Should we also invoke the onUndeliveredElement lambda in the order of elements addition?

No hard opinion here. My educated guess would be "yes, we should" if it doesn't drastically complicate the implementation

Reply:
I have changed the logic for resuming both receivers and senders in their order of suspension in the channel and added a test with suspended senders. 

I haven't touched the `onUndeliveredElement` logic as we discussed personally.